benchexec/container.py

```diff
@@ -723,28 +765,133 @@ def make_overlay_mount(mount, lower, upper, work):
+    by determining if there are any sub-mounts below it.
+    """
+    mount_points = [
+        (full_mountpoint, fstype)
+        for _unused_source, full_mountpoint, fstype, _options in get_mount_points()
+        if util.path_is_below(full_mountpoint, mount_base)
+    ]
+
+    for full_mountpoint, fstype in mount_points:
+        mountpoint = full_mountpoint[len(mount_base) :] or b"/"
+        mode = determine_directory_mode(dir_modes, mountpoint, fstype)
+
+        if not mode or not os.path.exists(mountpoint):
+            continue
+
+        if mode == DIR_OVERLAY:
+            # Check if there are any sub-mounts within the current overlay mount point
+            for sub_mountpoint, _unused_fstype in mount_points:
+                if (
+                    util.path_is_below(sub_mountpoint, mountpoint)
+                    and sub_mountpoint != mountpoint
+                ):
+                    return True
+
+    return False
+
+
+@contextlib.contextmanager
+def permitted_cap_as_ambient():
+    """
+    Transfer all permitted capabilities to the inheritable set
+    and raise them in the ambient set if effective.
+    Finanlly drop all ambient capabilities by removing them from the ambient set,
+    and undo changes made to inheritable set.
+
+    Used by fuse-based overlay mounts needing temporary capability elevation.
+    """
+    header = libc.CapHeader(libc.LINUX_CAPABILITY_VERSION_3, 0)
+    data = (libc.CapData * libc.LINUX_CAPABILITY_U32S_3)()
+
+    libc.capget(header, data)
+    original_inheritable = [data[0].inheritable, data[1].inheritable]
+    cap_last_cap = int(util.try_read_file("/proc/sys/kernel/cap_last_cap") or "0")
+
+    try:
+        data[0].inheritable = data[0].permitted
+        data[1].inheritable = data[1].permitted
+        libc.capset(header, data)
+
+        effective = (data[1].effective << 32) | data[0].effective
+        for cap in range(cap_last_cap + 1):
+            if effective & (1 << cap):
+                libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_RAISE, cap, 0, 0)
+
+        yield
+    finally:
+        libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_CLEAR_ALL, 0, 0, 0)
+
+        data[0].inheritable, data[1].inheritable = original_inheritable
+        libc.capset(header, data)
+
+
+def setup_fuse_overlay(temp_base, work_base):
+    """
+    Check if fuse-overlayfs is available on the system and,
+    if so, creates a temporary overlay filesystem by stacking the root directory
+    with a specified temporary directory.
+
+    @return: The path to the mounted overlay filesystem if successful, None otherwise.
+    """
+    fuse = shutil.which("fuse-overlayfs")
+    if fuse is None:
+        return None
+    temp_fuse = temp_base + b"/fuse"
+    work_fuse = work_base + b"/0"
+    os.makedirs(temp_fuse, exist_ok=True)
+    os.makedirs(work_fuse, exist_ok=True)
+
+    logging.debug(
+        "Creating overlay mount with fuse-overlayfs: target=%s, lower=%s, upper=%s, work=%s",
```

Comment:
We could log the full path in `fuse` here as debugging aid, e.g., by replacing `fuse-overlayfs` in the log message with it.